### PR TITLE
Update sampling methods and kv cache.

### PR DESCRIPTION
**Pull request by phkhanhtrinh23**
# Notes
Contributor: [phkhanhtrinh23](https://github.com/phkhanhtrinh23).

## KV Cache
It seems that your repository has not implemented the `kv_cache` for inference phases. Therefore, I have tried my best to implement it, based on this [reference](https://kipp.ly/transformer-inference-arithmetic/#kv-cache). It is crucial that we should not use the attention [function](https://github.com/karpathy/nanoGPT/blob/eba36e84649f3c6d840a93092cb779a260544d08/model.py#L64)  from PyTorch in this implementation. I have to use the manual implementation of attention function to run this, instead.

## Sampling methods (temperature, top-k, and top-p)
To increase the likelihood of high probability words and decreasing the likelihood of low probability words, we can apply the so-called `temperature` to the Softmax function. Moreover, I have also added the `top_k` and `top_p` sampling based on this [blog](https://huggingface.co/blog/how-to-generate#sampling). Although the original [code](https://github.com/karpathy/nanoGPT/blob/eba36e84649f3c6d840a93092cb779a260544d08/model.py#L306) have added the `temperature` and `top_k`, it does not have the implementation of `top_p`. Therefore, I have added all of them in this version.

## Contact
Thank you for your contribution! If you have any concerns, you can reach me at phkhanhtrinh23@gmail.com.

### CODE REVIEW

**Review by cin-toy — COMMENTED** [review]
I have read through your code regarding `kv_cache` and sampling methods. It is great to see your contributions. I appreciate it a lot!

I confirm that your code does not have any conflicts with mine. Therefore, I approve your contributions and dedication. Keep moving forward and devote yourself to the AI Community!
